Feature/add rest endpoints for product service

MSurmach opened the pull request.

MSurmach committed:
Add rest endpoints for product service

MSurmach committed:
Add product-service events

Tyomich4736 commented on product-service/src/main/java/com/intexsoft/webshop/productservice/controller/SubcategoryController.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/category/{categoryId}/subcategory")
```
categoryId тут не нужно, т.к. оно используется не во всех методах

MSurmach replied:
Разбиты на отдельные контроллеры /category / subcategory

Tyomich4736 commented on product-service/src/main/java/com/intexsoft/webshop/productservice/controller/SubcategoryController.java:
```diff
@@ -0,0 +1,56 @@
+
+    @PostMapping
+    public ResponseEntity<SubcategoryDto> createSubcategory(@RequestBody @Valid SubcategoryCreateDto subcategoryCreateDto,
+                                                            @PathVariable @Positive Long categoryId) {
```
Лучше сделать categoryId частью SubcategoryCreateDto

MSurmach replied:
сделано

Tyomich4736 commented on product-service/src/main/java/com/intexsoft/webshop/productservice/mapper/AttributeValueMapper.java:
```diff
@@ -0,0 +1,31 @@
+    default AttributeValue toAttributeValue(ProductAttributeValueCreateDto productAttributeValueCreateDto) {
+        AttributeValue attributeValue = new AttributeValue();
+        attributeValue.setValue(productAttributeValueCreateDto.getValue());
+        Attribute attribute = new Attribute();
```
То есть на каждое новое AttributeValue в базе будет создаваться новый Attribute?

MSurmach replied:
Логика создания и обновления продукта отрефакторена

MSurmach committed:
Add changes due to code review
- changes in endpoints (/category and /subcategory)
- logs
- shop_replica and product_shop_link tables were deleted
- shop's event consumer also was deleted
- refactoring and improvements

MSurmach committed:
Work with attributes and values

MSurmach committed:
Refactor product-service
- add separate exceptions
- fixes in update and create product logic

Tyomich4736 commented on product-service/src/main/java/com/intexsoft/webshop/productservice/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,206 @@
+
+    private List<AttributeValue> createAttributeValues(ProductCreateDto productCreateDto) {
+        List<AttributeValueCreateDto> attributeValueCreateDtos = productCreateDto.getAttributeValueCreateDtos();
+        if (isNull(attributeValueCreateDtos)) return null;
```
Хорошим тоном считается вызывать статические методы через имя их класса, а не с import static

MSurmach replied:
убрал статические импорты из этого класса

Tyomich4736 commented on product-service/src/main/java/com/intexsoft/webshop/productservice/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,206 @@
+                productForImagesUpdate.addImage(imageMapper.toImage(imageUpdateDto));
+                return;
+            }
+            Image image = product_idImage_map.get(imageId);
```
Используй строго camelCase для именования переменных

MSurmach replied:
Переименовано в camelCase

Tyomich4736 commented on product-service/src/main/java/com/intexsoft/webshop/productservice/exception/notfound404/AttributeNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+
+public class AttributeNotFoundException extends ResourceNotFoundException {
+    public AttributeNotFoundException(Long attributeId) {
+        super(ATTRIBUTE, attributeId);
```
В идеале тоже использовать элемент enum через его имя

MSurmach replied:
Поправил импорты. Вызов enum через класс

Tyomich4736 commented on product-service/src/main/java/com/intexsoft/webshop/productservice/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,206 @@
+                .map(attributeValueCreateDto -> {
+                    Attribute attribute = new Attribute();
+                    attribute.setId(attributeValueCreateDto.getAttributeId());
+                    attribute.setLabel(attributeIdLabelMap.get(attributeValueCreateDto.getAttributeId()));
```
Лишняя работа. В идеале лучше создавать новую сущность через new когда мы действительно хотим создать новую сущность в базе. Тогда не будет визуальной путаницы. Здесь выше можно создать мапу атрибутов по id (Map<Long, Attribute>) и при создании AttributeValu просто засетать атрибут из мапы

MSurmach committed:
Refactor due to code review
